tests/test_bootstrap.py

```diff
@@ -0,0 +1,551 @@
+import os
+
+import unittest
+
+try:
+    from unittest import mock
+except ImportError:
+    # `Python 2` or lower than `Python 3.3` does not
+    # have the `unittest.mock` module built-in
+    import mock
+from pythonforandroid.bootstrap import Bootstrap
+from pythonforandroid.distribution import Distribution
+from pythonforandroid.recipe import Recipe
+from pythonforandroid.archs import ArchARMv7_a
+from pythonforandroid.build import Context
+
+
+class BaseClassSetupBootstrap(object):
+    """
+    An class object which is intended to be used as a base class to configure
+    an inherited class of `unittest.TestCase`. This class will override the
+    `setUp` and `tearDown` methods.
+    """
+
+    def setUp(self):
+        self.ctx = Context()
+        self.ctx.ndk_api = 21
+        self.ctx.android_api = 27
+        self.ctx._sdk_dir = "/opt/android/android-sdk"
+        self.ctx._ndk_dir = "/opt/android/android-ndk"
+        self.ctx.setup_dirs(os.getcwd())
+        self.ctx.recipe_build_order = [
+            "hostpython3",
+            "python3",
+            "sdl2",
+            "kivy",
+        ]
+
+    def setUp_distribution_with_bootstrap(self, bs):
+        """
+        Extend the setUp by configuring a distribution, because some test
+        needs a distribution to be set to be properly tested
+        """
+        self.ctx.bootstrap = bs
+        self.ctx.bootstrap.distribution = Distribution.get_distribution(
+            self.ctx, name="test_prj", recipes=["python3", "kivy"]
+        )
+
+    def tearDown(self):
+        """
+        Extend the `tearDown` by configuring a distribution, because some test
+        needs a distribution to be set to be properly tested
+
+        .. note:: Here we need to reset `self.ctx.bootstrap.distribution`
+                  because, otherwise we will be unable to successfully pass the
+                  test :meth:`TestBootstrapBasic.test_attributes`
+        """
+        if self.ctx.bootstrap:
+            self.ctx.bootstrap.distribution = None
+        self.ctx.bootstrap = None
+
+
+class TestBootstrapBasic(BaseClassSetupBootstrap, unittest.TestCase):
+    """
+    An inherited class of `BaseClassSetupBootstrap` and `unittest.TestCase`
+    which will be used to perform tests for the methods/attributes shared
+    between all bootstraps which inherits from class
+    :class:`~pythonforandroid.bootstrap.Bootstrap`.
+    """
+
+    def test_properties(self):
+        """A test which will initialize a bootstrap and will check if the
+        values are the expected.
+        """
+        bs = Bootstrap().get_bootstrap("sdl2", self.ctx)
+        self.assertEqual(bs.name, "sdl2")
+        self.assertEqual(bs.jni_dir, "sdl2/jni")
+        self.assertEqual(bs.get_build_dir_name(), "sdl2-python3")
+
+        # test dist_dir error
+        with self.assertRaises(SystemExit) as e:
+            getattr(bs, "dist_dir")
```

Comment:
Minor: just wondering why don't we simply `bs.dist_dir`?

Reply:
aaa, this is because I have a linting running in background inside my `ide` and it marks your solution as an operation that does nothing...it seems that is happy with `getattr`...the final result is the same...but...since it seems that you have seen something strange in here, I have changed it as you proposed :wink:

Reply:
oh OK I was also wondering if it was maybe for the linting, thanks for clarifying